src/NuGet.Clients/NuGet.VisualStudio.Common/Telemetry/VsIntanceTelemetryEmit.cs

```diff
@@ -0,0 +1,203 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.ComponentModel.Composition;
+using System.Linq;
+using NuGet.Common;
+
+namespace NuGet.VisualStudio.Telemetry
+{
+    [Export(typeof(INuGetTelemetryCollector))]
+    [Export(typeof(VsIntanceTelemetryEmit))]
+    [PartCreationPolicy(CreationPolicy.Shared)]
+    public sealed class VsIntanceTelemetryEmit : VsInstanceTelemetryConsts, INuGetTelemetryCollector
+    {
+        // _solutionTelemetryEvents hold telemetry for current VS solution session.
+        private List<TelemetryEvent> _vsSolutionTelemetryEvents;
+        private Dictionary<string, object> _vsSolutionTelemetryEmitQueue;
+        // _vsInstanceTelemetryEvents hold telemetry for current VS instance session.
+        private List<TelemetryEvent> _vsInstanceTelemetryEvents;
+        private Dictionary<string, object> _vsInstanceTelemetryEmitQueue;
+
+        private int _solutionCount;
+
+        public VsIntanceTelemetryEmit()
+        {
+            _vsSolutionTelemetryEvents = new List<TelemetryEvent>();
+            _vsSolutionTelemetryEmitQueue = new Dictionary<string, object>();
+            _vsInstanceTelemetryEvents = new List<TelemetryEvent>();
+            _vsInstanceTelemetryEmitQueue = new Dictionary<string, object>();
+        }
+
+        // Adds telemetry into list which will be aggregated by end of VS solution sessions or VS instance session.
+        public void AddSolutionTelemetryEvent(TelemetryEvent telemetryData)
```

Comment:
Passing around telemetry object and populating them with additional data feels like we're not really separating concerns. 

There are some components that know about the data, others that know how/when to send telemetry. Now all of those components are passing around telemetry objects. 

You should consider separating the collecting of data from telemetry sending.

Reply:
I changed to Dictionary<string, object> type for telemetry collector, trying to keep it generic as possible.